### DISCUSSION TITLE
Gamepad emulation

### DISCUSSION THREAD

**Discussion by zehthales** [sezanzeb/input-remapper#464] [Q&A]
Whenever I apply the settings for my USB keyboard Steam says "Xbox360 controller connected". So I was wondering, it's possible to emulate Xbox360 from my USB keyboard? I mean, keyboard to gamepad emulation 

If don't, you guys have the potential of being the firsts to do this, since all gamepad mappers on Linux only do the gamepad to keyboard emulation, never the opposite.

**Comment by jonasBoss** [comment]
It is. 

You can use your keyboard to press all the different buttons on the virtual gamepad. Just select `gamepad` from the dropdown menue.

If you also want to use the axis (sticks) of the virtual gamepad you can write a macro with the `event()` function. 
`event(3, 0, 32768)` would set the left stick all the way to the left (or right?).

coming soon (currently in this PR: https://github.com/sezanzeb/input-remapper/pull/442) you will be able to use your mouse movement to move the sticks of the virtual gamepad.

**Reply by zehthales** [reply]
I see. Thanks for the reply. I tried it now and nothing happened. There is any other configuration to after applying the change?

**Reply by jonasBoss** [reply]
What exactly did you try? can you share the json file of the mapping?

**Reply by zehthales** [reply]
With input-remapper 1.5 recorded the key(q), selected gampad for the output and wrote event(3, 0, 32768). Also tried the dpad: event(1, 16, -1)

**Reply by jonasBoss** [reply]
Can you check with `$sudo evtest` that the injection for `input-remapper gamepad` is actually not working. 
Also `event(3, 0, 32768)` will only push the stick to the side, but never release it. Try this instead: `hold(event(3,0,32768)).event(3,0,0)`.

The d-pad is actually also two axis with the codes `ABS_HAT0X` and `ABS_HAT0Y` so you will need `event(3,10,-1)` or `event(3,11,-1)`

**Reply by felgon** [reply]
> Try this instead: `hold(event(3,0,32768)).event(3,0,0)`.

Can confirm that this template works, thank you very much!
(tested on version 2.2.0)

Edit: note that holding down multiple keys mapped to the same event code seems to result in a race condition e.g. simultaneously holding down both left and right movement keys seems to result in a final input that keeps randomly alternating between left and right movement.

**Comment by zehthales** [comment]
This macro hold(event(3,0,32768)).event(3,0,0) made my USB keyboard shut down somehow. Same thing for axis. But it really works for buttons, which is the only thing I don't need remapping

**Reply by jonasBoss** [reply]
That is strange. Maybe the problem is that `hold()` actually repeats a macro until the key is released. Can you try `event(3,0,32768).hold().event(3,0,0)` instead?

**Reply by zehthales** [reply]
it really worked with analogs and axis (something like this event(3,17,-1).hold().event(3,17,0)). Now, i need to see if the steam game will get the input

**Reply by zehthales** [reply]
Now will test if the mapping works inside the game and what is the rule for left+right.

**Reply by zehthales** [reply]
Test results:
Note: the mapping to gamepad worked inside the game

Mapping-1:  d = event(3,0,32768).hold().event(3,0,0) and a = event(3,0,-32768).hold().event(3,0,0)
-> The event(3,0,0) is affecting the simultaneous press of both key, pressing a+d as and releasing a, the event(3,0,0) is sent and the character just stand at the spot. Same thing for d+a. 

Mapping-2: d = event(3,16,-1).hold().event(3,16,0) and a = event(3,16,1).hold().event(3,16,0)
Same as the mapping-1

Mapping 3: d = event(3,0,32768).hold().event(3,0,0) and a = event(3,16,1).hold().event(3,16,0)
Works fine, but the rules of left+right is the same as the steam keyboad mapping

**Reply by felgon** [reply]
> That is strange. Maybe the problem is that `hold()` actually repeats a macro until the key is released. Can you try `event(3,0,32768).hold().event(3,0,0)` instead?

This template works as well, though be aware that this makes holding down multiple keys mapped to the same event code followed by releasing one of the keys result in a neutral input e.g. holding down both `left` and `right` movement keys and then releasing `left` will make your character stop moving instead of moving `right`.
(tested on version 2.2.0)
